### PR TITLE
Cleaner GRANDPA RPC API for proving finality

### PR DESCRIPTION
**Pull request by octol**
Implements #7115 as in the description, with the exception that the RPC call takes block numbers instead of block hashes

polkadot companion: https://github.com/paritytech/polkadot/pull/2100

TODO:
- [X] ~~Store block numbers in aux_schema~~
- [X] Rewrite to store block numbers as part of `AuthoritySet` instead
- [X] Manually implement `Decode` for `AuthoritySet` for compatibility
- [x] Flesh out all WIPs and handle all unwraps
- [x] Clear out `finality_proof.rs` now that the light client isn't using it anymore
- [x] ~~Fix: prove finality with the last block of the set fails~~
- [x] Tests
- [x] Handle the case when the stored indices is incomplete (missing beginning)
- [x] What are the implications of forced changes?
- [x] Should we replace `FinalityProofFragment` with something more tailored for this use case?

### CODE REVIEW

File: client/finality-grandpa/src/authority_set_changes.rs
Comment (andresilva) [inline client/finality-grandpa/src/authority_set_changes.rs:41]: I know I suggested not storing the set_id and instead implicitly figuring out by the position in the array. The only disadvantage of this approach is that it breaks for nodes that didn't sync from scratch with this code. I guess it might be better to store the `set_id` explicitly. Additionally in the future we might start syncing from a specific checkpoint (rather from genesis always) or we may warp sync, in both cases we won't import all blocks.

File: client/finality-grandpa/src/authority_set_changes.rs
Comment (andresilva) [inline client/finality-grandpa/src/authority_set_changes.rs:23]: Maybe this can go inside `AuthoritySet`? We will always be operating on both at the same time, i.e. when we finalize something from `AuthoritySet` is also when we update this.

**Review by andresilva — COMMENTED** [review]
Looks to be going in the right direction :)

**Comment by octol** [comment]
Think this is probably ready for another round of feedback now, and finished enough to be upgraded from draft to PR

Most things have been reworked according to feedback from previous versions, as well as #7546 landing in master gives us a lot more freedom in changing `finality_proofs.rs`

**Review by wheresaddie — APPROVED** [review]
lgtm ty

File: client/finality-grandpa/src/authorities.rs
Comment (octol) [inline client/finality-grandpa/src/authorities.rs:693]: I'm sort of thinking about storing this as `(set_id, (first_block, last_block))` to make the consistency checks easier for when the set of changes is incomplete

File: client/finality-grandpa/rpc/src/finality.rs
Comment (andresilva) [inline client/finality-grandpa/rpc/src/finality.rs:29]: I guess we can copy the docs from `prove_finality`.

File: client/finality-grandpa/rpc/src/lib.rs
Comment (andresilva) [inline client/finality-grandpa/rpc/src/lib.rs:51]: Let's name it `Number`, I've taken some fondness to descriptive type parameter names and it's more inline with the style of the existing parameters.

File: client/finality-grandpa/rpc/src/lib.rs
Comment (andresilva) [inline client/finality-grandpa/rpc/src/lib.rs:86]: We also provide all the intermediary headers :)

File: client/finality-grandpa/rpc/src/finality.rs
Comment (andresilva) [inline client/finality-grandpa/rpc/src/finality.rs:26]: Maybe it makes sense to rename `EncodedFinalityProofs` to `EncodedFinalityProof` since now it's only one and not multiple fragments.

File: client/finality-grandpa/rpc/src/lib.rs
Comment (andresilva) [inline client/finality-grandpa/rpc/src/lib.rs:536]: Let's just name this `finality_proof_rpc` or something like that.

File: client/finality-grandpa/src/authorities.rs
Comment (andresilva) [inline client/finality-grandpa/src/authorities.rs:170]:
```suggestion
	/// Track at which blocks the set id changed. This is useful when we need to prove finality for a given block
	/// since we can figure out what set the block belongs to and when the set started/ended.
```

File: client/finality-grandpa/src/authorities.rs
Comment (andresilva) [inline client/finality-grandpa/src/authorities.rs:444]: We should keep this immutable.

File: client/finality-grandpa/src/authorities.rs
Comment (andresilva) [inline client/finality-grandpa/src/authorities.rs:506]: This method does not mutate and instead returns a new `AuthoritySet` if there are any changes. This is how we handle rollbacks in case we apply a change and later on in the process the block being imported fails for some reason. Instead we should keep this method immutable and we should instead create the appended version of `AuthoritySetChanges` and use it below.

File: client/finality-grandpa/src/authorities.rs
Comment (andresilva) [inline client/finality-grandpa/src/authorities.rs:689]: We should make it clear here that this tracks historical authority set changes that have already happened. Not to be confused with the rest of the fields above where we are tracking possible upcoming authority set changes.

File: client/finality-grandpa/src/authorities.rs
Comment (andresilva) [inline client/finality-grandpa/src/authorities.rs:708]: Bad indentation.

File: client/finality-grandpa/src/authorities.rs
Comment (andresilva) [inline client/finality-grandpa/src/authorities.rs:709]:
```suggestion
			.binary_search_by_key(&block_number, |(_, n)| n.clone())
```

File: client/finality-grandpa/src/authorities.rs
Comment (andresilva) [inline client/finality-grandpa/src/authorities.rs:719]: Maybe we can one of the tests that apply standard/forced changes to assert that the authority set change is properly tracked.

File: client/finality-grandpa/src/finality_proof.rs
Comment (andresilva) [inline client/finality-grandpa/src/finality_proof.rs:59]: Do we still need this trait and impls?

File: client/finality-grandpa/src/finality_proof.rs
Comment (andresilva) [inline client/finality-grandpa/src/finality_proof.rs:65]: Same as above.

File: client/finality-grandpa/src/finality_proof.rs
Comment (andresilva) [inline client/finality-grandpa/src/finality_proof.rs:149]: The only downside of the current implementation is that we can't prove finality of blocks in the current set.

File: client/finality-grandpa/src/authorities.rs
Comment (andresilva) [reply client/finality-grandpa/src/authorities.rs:693]: I guess we could just check that this vec is missing entries for older set ids to figure out that we don't have all data. E.g. if we have set id 3 in there then we should have all the ones before that as well contiguosly.

**Review by andresilva — CHANGES_REQUESTED** [review]
A bunch of comments but overall the changes look good :)

File: client/finality-grandpa/src/finality_proof.rs
Comment (octol) [reply client/finality-grandpa/src/finality_proof.rs:139]: Agree, a bit unfortunate. Do you think it's worth adding additional code paths for this purpose?

**Comment by octol** [comment]
Significant rework following the reviews, hopefully we should be a lot closer to the finishing line now :)

File: client/finality-grandpa/src/finality_proof.rs
Comment (octol) [reply client/finality-grandpa/src/finality_proof.rs:62]: I've removed  `StorageAndProofProvider`, but kept `AuthoritySetForFinalityProver` since it's useful for the tests

File: client/finality-grandpa/src/finality_proof.rs
Comment (octol) [reply client/finality-grandpa/src/finality_proof.rs:139]: I've added a return message to suggest using the subscription API in this case for now

**Comment by gnunicorn** [comment]
@octol what's the status?

**Comment by octol** [comment]
> @octol what's the status?

I've just been working on updating this to latest master, and @andresilva kindly agreed to have another look

File: client/finality-grandpa/src/finality_proof.rs
Comment (andresilva) [inline client/finality-grandpa/src/finality_proof.rs:174]: I think this shouldn't be needed, this is why I asked above if this trait is still needed. I think this is just an hold over from the previous implementation where we didn't know for sure when a set started or ended, therefore we figured it out by iterating over blocks and checking whether the authorities changed. Now we know for sure the last block in set X is block Y, so we can just fetch it from the database and extract the justification from it without doing any further checking. The current implementation is looking into runtime state at a given block to figure authorities, which means it won't work for old blocks if state is pruned. 

Either way I won't block this PR on this. We can deal with these changes in a follow-up PR.

File: primitives/blockchain/src/error.rs
Comment (andresilva) [reply primitives/blockchain/src/error.rs:104]: It seems like it's unused in the PR but I may have overlooked.

**Review by andresilva — CHANGES_REQUESTED** [review]
Just small nits now

File: primitives/blockchain/src/error.rs
Comment (octol) [reply primitives/blockchain/src/error.rs:104]: Currently it's used in `fn prove_finality()` to signal to the user e.g. to try the pubsub API if a block in the latest set_id is requested

I'll see if I can move it somewhere else

Comment (octol) [reply primitives/blockchain/src/error.rs]: Updated the PR with a new error enum specific to finality_proof

**Review by andresilva — APPROVED** [review]

**Comment by andresilva** [comment]
bot merge

File: client/finality-grandpa/src/finality_proof.rs
Comment (octol) [reply client/finality-grandpa/src/finality_proof.rs:190]: Created a new branch that I can raise as a follow-up PR after this is merged

File: client/finality-grandpa/src/finality_proof.rs
Comment (bkchr) [inline client/finality-grandpa/src/finality_proof.rs:266]:
```suggestion
					Returning empty proof",
```

File: client/finality-grandpa/src/authorities.rs
Comment (bkchr) [inline client/finality-grandpa/src/authorities.rs:186]: Where is this decode being used @andresilva ?

If someone uses this where there is still data in `I`, this could lead to decoding shit :P

File: client/finality-grandpa/src/authorities.rs
Comment (bkchr) [inline client/finality-grandpa/src/authorities.rs:693]:
```suggestion
// authority set, once they have been finalized.
```

File: client/finality-grandpa/src/authorities.rs
Comment (bkchr) [inline client/finality-grandpa/src/authorities.rs:715]: How can that fail? Wouldn't that mean that the entire object is wrong?

**Review by bkchr — APPROVED** [review]
LGTM as far as I understand it :P

File: client/finality-grandpa/src/authorities.rs
Comment (octol) [reply client/finality-grandpa/src/authorities.rs:715]: Yeah pretty much, if this happens the object is missing authority set changes and something must have gone wrong applying changes. It's my understanding that this can't happen, but I'm not certain enough of that to omit this check ... @andresilva what do you think?

Comment (andresilva) [reply client/finality-grandpa/src/authorities.rs:715]: We wanted to fail gracefully for nodes that will migrate to this and therefore won't have all set ids stored here (only when they sync from scratch), the only check that's really needed for this is the first one `idx < self.0.len()`. These checks don't hurt but they can be removed since they would imply that this data structure skipped set ids spuriously which is wrong.

File: client/finality-grandpa/src/authorities.rs
Comment (andresilva) [reply client/finality-grandpa/src/authorities.rs:186]: Currently it's only used through this https://github.com/paritytech/substrate/blob/master/client/finality-grandpa/src/aux_schema.rs#L110, so I don't think it is ever decoded as part of a bigger structure (i.e. there is no more data to decode). It would be nice that we don't have a footgun here though, do you have any suggestion?

Comment (bkchr) [reply client/finality-grandpa/src/authorities.rs:186]: Maybe we should offer some migration? So, that we only apply this once? But I assume we don't have any way to store a version in aux?

Comment (bkchr) [reply client/finality-grandpa/src/authorities.rs:186]: Besides that, I don't have better ideas.

Comment (andresilva) [reply client/finality-grandpa/src/authorities.rs:186]: I was already not remembering the `aux_schema` code for GRANDPA. We indeed do have a version there and we have code there that handles migrations of GRANDPA data. So yeah, we should definitely handle this the same way.

**Review by andresilva — CHANGES_REQUESTED** [review]
@octol `LegacyAuthoritySet` should be moved to `aux_schema.rs` and renamed to `V2AuthoritySet`. The `CURRENT_VERSION` const there should be incremented and a method `migration_from_version2` should be added (and then called from `load_persistent`). This method will read all remaining GRANDPA structures as they are but decode the authority set as `V2AuthoritySet` and then return the regular `AuthoritySet` (with `AuthoritySetChanges::empty()`).

**Comment by octol** [comment]
> @octol `LegacyAuthoritySet` should be moved to `aux_schema.rs` and renamed to `V2AuthoritySet`. The `CURRENT_VERSION` const there should be incremented and a method `migration_from_version2` should be added (and then called from `load_persistent`). This method will read all remaining GRANDPA structures as they are but decode the authority set as `V2AuthoritySet` and then return the regular `AuthoritySet` (with `AuthoritySetChanges::empty()`).

Oh thanks, yeah that will make it much better!

**Comment by octol** [comment]
Pushed a new commit that implements the migration logic as suggested by @andresilva

**Review by andresilva — APPROVED** [review]
LGTM, thanks!

**Comment by andresilva** [comment]
There were conflicts in `finality_proof.rs` because of #7711. Merged master and fixed them.

**Comment by andresilva** [comment]
bot merge

**Comment by octol** [comment]
bot merge

**Pull request merged**
